### PR TITLE
Add support for CHaP and ghc-9.2

### PR DESCRIPTION
**Pull request by erikd**
Builds with both ghc-8.10 and ghc-9.2.

Currently uses a `source-package-repository` reference to a `cardano-ledger` PR that has not yet been merged.

### CODE REVIEW

**Comment by angerman** [comment]
The windows failures might partially be due to https://github.com/msys2/MINGW-packages/issues/10837#issuecomment-1047105402; more investigation needed.

File: ouroboros-network-framework/src/Ouroboros/Network/ErrorPolicy.hs
Comment (angerman) [reply ouroboros-network-framework/src/Ouroboros/Network/ErrorPolicy.hs:255]: you don't like my ascii facepalm?

Comment (coot) [reply ouroboros-network-framework/src/Ouroboros/Network/ErrorPolicy.hs:255]: oh I see ok, I didn't know that's how facepalm looks like in ascii.

File: ouroboros-network/src/Ouroboros/Network/Protocol/LocalTxMonitor/Type.hs
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Protocol/LocalTxMonitor/Type.hs:227]: let's keep the code aligned
```suggestion
  exclusionLemma_ClientAndServerHaveAgency TokIdle     tok = case tok of {}

  exclusionLemma_ClientAndServerHaveAgency TokAcquired tok = case tok of {}

  exclusionLemma_NobodyAndClientHaveAgency TokDone     tok = case tok of {}
```

File: cabal.project
Comment (coot) [inline cabal.project:132]: @dnadales you should distribute these constraints in `ouroboros-consensus` packges.

@angerman I'd prefer `^>=` so we don't get CI build failures as soon as a new version of `cardano-ledger` or `plutus` is published to `CHaP.

**Review by coot — CHANGES_REQUESTED** [review]

**Comment by angerman** [comment]
bors merge

**Review by coot — APPROVED** [review]
LGTM

**Comment by angerman** [comment]
bors merge

**Comment by amesgen** [comment]
@angerman Is it intentional that we don't test with GHC 9.2 on Hydra/is there are a follow-up PR already planned? We don't run the Consensus tests on GH Actions, so this means that GHC 9.2 is untested on CI for Consensus.

File: ouroboros-consensus-shelley-test/src/Test/ThreadNet/TxGen/Shelley.hs
Comment (amesgen) [inline ouroboros-consensus-shelley-test/src/Test/ThreadNet/TxGen/Shelley.hs:66]: This is a logic change, isn't it?

**Review by amesgen — CHANGES_REQUESTED** [review]
Usually, PRs that touch both Network and Consensus should be reviewed by team members from both Consensus and Network (see also the codeowners) -- maybe we should make that more clear.

**Comment by amesgen** [comment]
bors cancel

File: ouroboros-consensus-shelley-test/src/Test/ThreadNet/TxGen/Shelley.hs
Comment (erikd) [reply ouroboros-consensus-shelley-test/src/Test/ThreadNet/TxGen/Shelley.hs:66]: Furthermore, if the code above the deleted code  did not cover **all** cases, there would have been a compiler warning and there wasn't.

I also find it useful to convert `-Wall` warnings to errors using `-Werror` . I trust GHC far more than I trust myself.

Comment (amesgen) [reply ouroboros-consensus-shelley-test/src/Test/ThreadNet/TxGen/Shelley.hs:66]: The problem is not that not all cases all covered, but rather that there is a logic change 

The code in question looked like this:

```haskell
  testGenTxs _coreNodeId _numCoreNodes curSlotNo cfg extra lst
      | stgeStartAt > curSlotNo = pure []

      -- TODO Temporarily disable the transaction generator until we fix the
      -- failing assertion in TxSubmission.Inbound, see #2680.
      --
      -- When fixed, remove the line below to re-enable the transaction
      -- generator.
      | otherwise               = pure []

      | otherwise               = do
      n <- choose (0, 20)
      go [] n $ applyChainTick lcfg curSlotNo lst
```

The first `| otherwise` clause was intentionally added in order to work around an existing bug (#2680). Now the pattern match redundancy checker got smarter in GHC 9.2 compared to 8.10, and detects that the second `| otherwise` clause is unreachable (which is intentional here!).

To fix this, you can use a more "complicated" expression that evaluates to `True` instead of `otherwise`, e.g. `() == ()` (or even the new [`considerAccessible`](https://hackage.haskell.org/package/base-4.17.0.0/docs/GHC-Exts.html#v:considerAccessible) that has exactly this purpose).

Does that clear it up?

---

> I also find it useful to convert `-Wall` warnings to errors using `-Werror`.

Agreed, we do that on Hydra, but this PR does not add 9.2 to Hydra CI (see https://github.com/input-output-hk/ouroboros-network/pull/4144#issuecomment-1322006034).

Comment (erikd) [reply ouroboros-consensus-shelley-test/src/Test/ThreadNet/TxGen/Shelley.hs:66]: The current way the last case is disabled is confusing to people not a part of this project in that it current generates a compiler warning that someone not familiar with the code thinks they should remove.

**Comment by angerman** [comment]
@amesgen who can review this? It's overwhelming mechanical changes to add 9.2 compatibility, and it's blocking downstream work on node.

File: ouroboros-consensus-shelley-test/src/Test/ThreadNet/TxGen/Shelley.hs
Comment (erikd) [reply ouroboros-consensus-shelley-test/src/Test/ThreadNet/TxGen/Shelley.hs:66]: I have come up with an alternative that both maintains the logic and satisfies GHC.

Comment (amesgen) [reply ouroboros-consensus-shelley-test/src/Test/ThreadNet/TxGen/Shelley.hs:66]: Thanks, using an `if True` is another way to "outsmart" even the improved GHC 9.2 pattern match redundancy checker :smile:

**Comment by amesgen** [comment]
> who can review this? It's overwhelming mechanical changes to add 9.2 compatibility, and it's blocking downstream work on node.

I think the only thing missing now is GHC 9.2 on Nix/Hydra -- see https://github.com/input-output-hk/ouroboros-network/pull/4144#issuecomment-1322006034, and this is also the only place where we have `-Werror` enabled for the Consensus packages: While testing everything locally, I found a few warnings in the Consensus code (also pushed a commit fixing those, feel free to squash).

File: ouroboros-consensus-cardano-tools/app/DBAnalyser/Parsers.hs
Comment (nfrisby) [inline ouroboros-consensus-cardano-tools/app/DBAnalyser/Parsers.hs:6]: Could we avoid CPP here if we used a qualified import?

A very rushed experiment suggests that making this a qualified import and similarly qualifying the `asum` occurrences in this module (I assume there aren't many) would avoid the warning.

File: ouroboros-consensus-protocol/src/Ouroboros/Consensus/Protocol/Praos.hs
Comment (nfrisby) [inline ouroboros-consensus-protocol/src/Ouroboros/Consensus/Protocol/Praos.hs:365]: Happy to see this, thanks.

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs
Comment (nfrisby) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs:320]: I don't immediately recognize why this is necessary. (note to self)

File: ouroboros-consensus-test/src/Test/ThreadNet/Network.hs
Comment (nfrisby) [inline ouroboros-consensus-test/src/Test/ThreadNet/Network.hs:10]: I don't immediately recognize why this is necessary. (note to self)

File: ouroboros-consensus-test/src/Test/Util/TestBlock.hs
Comment (nfrisby) [inline ouroboros-consensus-test/src/Test/Util/TestBlock.hs:503]: I don't immediately recognize why this is necessary. (note to self)

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs
Comment (nfrisby) [inline ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs:663]: Is GHC 9.2 more sensitive about how the TH splices delineate the module's declarations now?

File: ouroboros-consensus/src/Ouroboros/Consensus/HeaderStateHistory.hs
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/HeaderStateHistory.hs:110]: Was there a bug in GHC <9.2 that caused it to be missing unnecessary constraints or something?

**Review by nfrisby — APPROVED** [review]
I Approve the Consensus changes.

There are several bits I don't already understand the reason for (see my comments), but they're all minor, so not blocking this PR. Thanks!

**Comment by nfrisby** [comment]
Just making sure GitHub notifies people about my edit just now to https://github.com/input-output-hk/ouroboros-network/pull/4144#pullrequestreview-1188175154 regarding Esgen's reviews.

File: ouroboros-consensus/src/Ouroboros/Consensus/HeaderStateHistory.hs
Comment (erikd) [reply ouroboros-consensus/src/Ouroboros/Consensus/HeaderStateHistory.hs:110]: Yes, in earlier GHCs that was incorrectly considered an necessary constraint. It never was necessary and with 9.2 it is flagged as redundant.

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs
Comment (erikd) [reply ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs:663]: Yes.

**Comment by erikd** [comment]
> Usually, PRs that touch both Network and Consensus should be reviewed by team members from both Consensus and Network (see also the codeowners) -- maybe we should make that more clear.

That should be resolved in the `CODEOWNERS` file. Please update it, or provide a diff on what it should be.

**Comment by angerman** [comment]
> > who can review this? It's overwhelming mechanical changes to add 9.2 compatibility, and it's blocking downstream work on node.
> 
> 
> 
> I think the only thing missing now is GHC 9.2 on Nix/Hydra -- see https://github.com/input-output-hk/ouroboros-network/pull/4144#issuecomment-1322006034, and this is also the only place where we have `-Werror` enabled for the Consensus packages: While testing everything locally, I found a few warnings in the Consensus code (also pushed a commit fixing those, feel free to squash).

As we are going to lose hydra in a week anyway, I don't think we should spend much time on hydra right now. There is a different Cicero PR I believe.

**Comment by angerman** [comment]
@amesgen, @nfrisby, @coot could we run those tests on GHA if we had our own runners?

**Review by amesgen — APPROVED** [review]
Thanks, LGTM!
  
> As we are going to lose hydra in a week anyway, I don't think we should spend much time on hydra right now. There is a different Cicero PR I believe.

Ok, I ran all tests locally using GHC 9.2, so I think it is fine to postpone it and include it in the mentioned #4108 :+1:

**Comment by amesgen** [comment]
> @amesgen, @nfrisby, @coot could we run those tests on GHA if we had our own runners?

If they were sufficiently performant, I think that would remove the reason why we currently don't do that. Note however that the Network tests *are* run on GHA even today, in contrast to the Consensus tests.

**Comment by amesgen** [comment]
bors retry

**Pull request merged**
